Feature/ncep bufr

markjolah opened the pull request:
This PR should unify the jedi-stack bufrlib implementations and make them interchangeable for the purposes of jedi-stack, and prepare for eventual switch to NCEPLIBS-bufr  See also: [jcsda/bufrlib # 11](https://github.com/JCSDA/bufrlib/pull/11)

The problem is that NCEPLIBS-bufr does not yet have shared library support, also there are some potential issues with setting fortran compiler flags.  Until these are worked out it may still be best to stick with the `jcsda/bufrlib` build.  However to switch implementations, all that needs to change here would be the commented-out line setting `software=NCEPLIBS-bufr`.

This PR and a rebuild of jedi-stack based container images must be coordinated with a merge of [ioda-converters # 311](https://github.com/JCSDA/ioda-converters/pull/311) which will adapt ioda-converters bufr based converters to use the new target names from NCEPLIBS-bufr.

markjolah commented:
A couple minor bugfixes.  I noticed pyncepbufr was still relying on `bufrlib_ROOT` based on my previous advice.  I changed the bufrlib module to export both `bufrlib_ROOT` and `bufr_ROOT`, but I will submit a subsequent PR to pyncepbufr to switch over to `bufr_ROOT` once we have switched to the new name in all environments.  Also build in release mode for optimizations.

markjolah commented:
OK.  Now with [`bufrlib` # 11](https://github.com/JCSDA/bufrlib/pull/11) merged, I made a release and named the tag the same in NCEPLIBS-bufr `bufr_v11.3.2`.  I tested and this stack build will also work fine changing to the comeneted out `software=NCEPLIBS-bufr`, but I did not actually have time to build a bundle and run tests.   If the current nceplibs-bufr is tested and working for everyone we can just go to there now editing that comment.

mmiesch commented:
Ok @markjolah - this is now installed on S4.  Use this stack:
```
module load jedi/intel-impi
```
i.e. without the "19" - `jedi/intel19-impi` is still the old one.  I'll update the docs to use the new one after we have tested it.  This is enabled cor C++14 and has the new buflib.   Let me know if you need anything else in the stack.

mmiesch commented on modulefiles/compiler/compilerName/compilerVersion/bufrlib/bufrlib.lua:
```diff
@@ -18,12 +18,9 @@ prepend_path("LD_LIBRARY_PATH", pathJoin(base,"lib"))
 prepend_path("DYLD_LIBRARY_PATH", pathJoin(base,"lib"))
 prepend_path("LIBRARY_PATH", pathJoin(base,"lib"))
 
```
Just one question - why get rid of these variables?  It can't hurt to keep some, can it?  For example, when I install module x, I often check the installation by going to directory `x_ROOT`.  So, it would be convenient for me to keep `bufrlib_ROOT` even though it's not strictly needed for the build system.    Similarly, the INCLUDES and LIBRARIES might be handy if you ever want to build something manually and link to the libraries.

markjolah replied:
Mainly as a simplification to prevent users from relying on them.  We were having general confusion about `<pkg>_PATH`, `<pkg>_ROOT`, and `<pkg>_DIR`.  With the new CMake versions we just need the `<pkg>_ROOT`.  With the new with NCEPLIBS-bufr naming schema `<pkg>` is now `bufr` instead of  `bufrlib`.  For backwards compatibility I added both `bufr_ROOT` and `bufrlib_ROOT` (in particular pyncepbufr was is still using `bufrlib_ROOT`).  If there is a compelling need for one of these other variable we can definitely add it back in, I just wanted to keep us focused on the minimum necessary.

markjolah replied:
I'm also trying to comment how the variables are getting used so when we review this a year from now we don't have the same questions about which variables are used where.

markjolah commented:
I found a bug with the way I setup the jcsda/bufrlib CMake configuration.  Hold off on merging this until I can address it.

markjolah commented:
@mmiesch The bug was a one-liner omission and is now fixed.  I updated the JCSDA/bufrlib release v11.3.2 and tag with the fix.  So if I could ask you to rebuild this one more time on S4 that would be great.  *NOTE:* you need to delete the `jedi-stack/pkg/bufrlib` checkout tempfile to make it reload the tags from github.  Sorry I should have tested this more thoroughly.

markjolah commented:
You can confirm the new version was installed correctly by looking for this line:
``` 
$ grep IMPORTED_GLOBAL ~/opt/modules/intel-20.0/bufrlib/11.3.2/lib/cmake/bufr/bufr-config.cmake
set_target_properties(${bufr_STATIC_LIBRARIES} ${bufr_SHARED_LIBRARIES} PROPERTIES IMPORTED_GLOBAL 1) #Allow Alias to targets
```

mmiesch commented:
Ok - it should be good to go on S4:
```
[mmiesch@s4-submit ~]$ grep IMPORTED_GLOBAL $bufr_ROOT/lib/cmake/bufr/bufr-config.cmake
set_target_properties(${bufr_STATIC_LIBRARIES} ${bufr_SHARED_LIBRARIES} PROPERTIES IMPORTED_GLOBAL 1) #Allow Alias to targets
```

markjolah commented:
Thanks @mmiesch.  I updated the bufrlib versions for the apps files in this PR, but I see you have already done it yourself on the actual file for S4.

mmiesch commented:
A strange problem installing this on Orion with the gnu-openmpi stack:
```bash

ifort: error #10052: could not checkout FLEXlm license

Error: A license for Comp-FL could not be obtained.  (-1,359,2).

Is your license file in the right location and readable?
The location of your license file should be specified via
the $INTEL_LICENSE_FILE environment variable.
```

Strange that it's trying to access an intel license even though I'm using the gnu compiler:
```bash
Currently Loaded Modules:
  1) git/2.28.0   2) python/3.7.5   3) gcc/8.3.0   4) jedi-gnu/8.3.0   5) cmake/3.18.1
```

Does this mean anything to you or is it some glitch on orion?

rickgrubin commented:
I bet that, just above those errors, you'll see that the Intel MPI compiler is being used, instead of `gfortran`; something like:

`/apps/intel-2020/intel-2020/compilers_and_libraries_2020.0.166/linux/bin/intel64/ifort [...]`

and the license referred to is what is loaded when the `module load license-intel` command is issued. Somewhere, the wrong compiler is getting loaded.

mmiesch commented:
@markjolah @rickgrubin should I go ahead and merge this?

mmiesch commented:
Looks like I haven't installed it on Cheyenne yet - I'll do so now.  I'll also update hdf5, netcdf, etc for gnu

mmiesch commented:
Ok - this new bufrlib is now part of the default `jedi/intel-impi` stack on Cheyenne.

But - a problem with gnu 9.1:

```
CMake Error at /glade/u/home/miesch/jedi/jedi-stack/pkg/bufrlib/build/CMakeFiles/_CMakeLTOTest-C/src/CMakeLists.txt:7 (add_executable):
  Compiler doesn't support IPO
```

I see they now have a gnu/10.1 module available - I can try rebuilding everything with that

mmiesch commented:
Same error with the `gnu/10.1.0` compiler:

```
-- Option: OPT_IPO: ON
CMake Error at /glade/u/home/miesch/jedi/jedi-stack/pkg/bufrlib/build/CMakeFiles/_CMakeLTOTest-C/src/CMakeLists.txt:7 (add_executable):
  Compiler doesn't support IPO


CMake Error at /glade/u/home/miesch/jedi/jedi-stack/pkg/bufrlib/build/CMakeFiles/_CMakeLTOTest-C/src/CMakeLists.txt:6 (add_library):
  Compiler doesn't support IPO

```

mmiesch commented:
Very strange on Orion with gnu (it works fine for intel).

It kept wanting to use ifort so I had to explictly tell it not to:

```bash
+ cmake -H. -Bbuild -DCMAKE_INSTALL_PREFIX=/work/noaa/da/mmiesch/modules/gnu-8.3.0/bufrlib/11.3.2 -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=1 -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc
```

That helped - it got it to compile - but now it's trying to install in /usr/local even though the above install prefix is clearly set correctly:

```
Install the project...
/apps/cmake-3.18.1/bin/cmake -P cmake_install.cmake
-- Install configuration: "Release"
-- Installing: /usr/local/lib64/bufr/module/GNU/8.3.0
CMake Error at cmake_install.cmake:46 (file):
  file INSTALL cannot make directory
  "/usr/local/lib64/bufr/module/GNU/8.3.0": No such file or directory.
```

mmiesch commented:
ok - not sure why - but it worked if I did this:
```
#VERBOSE=$MAKE_VERBOSE make -j${NTHREADS:-4}
#VERBOSE=$MAKE_VERBOSE $SUDO make install
make -j4
make VERBOSE=1 install
```
Now bufrlib/11.3.2 is included in both the intel and the gnu stacks on Orion - ready for testing @markjolah @rickgrubin
